### DISCUSSION TITLE
Portable support enhancement ideas

### DISCUSSION THREAD

**Discussion by villares** [tabreturn/thonny-py5mode#57] [Ideas]
I'm looking at this situation:
- I have downloaded Thonny 4 portable (with Python 3.10.6) and
- Installation of py5thonny mode worked fine on Windows 10.
- First activation of "imported mode for py5" downloaded the JDK-17 as expected, everything wonderful.

Then, when I move the portable folder arround, it breaks: The JDK is there in the right place, but the JAVA_HOME information in Thonny's configs/options got outdated.

It can be hard for beginners to understand what went wrong. If you uncheck and check the "imported mode for py5" menu a new updated JAVA_HOME line is added to the Thonny options, but the old one is not removed, and as far as I have tested this still won't work (you have to remove the outaded line, I think).

How could we improve this?

I had a proof of concept portable with everything already installed, but it counted on a laucher that would set JAVA-HOME every time, and also I had to modify thonny-py5mode so it would not add JAVA_HOME to Thonny options... I think this was far from ideal. So I'd like to enable people to build their own portables, and move them arround: that would be a good thing.

Is it too dangerous to check during plug-in loading if Thonny is portable and if so if JDK-17 is already installed and somehow act on it?  

I hate to add too much magic and to override something the user could change in the options panel manually but maybe for this kind of beginner friendly tool it's reasonable. I'm really not sure at all how we should proceed.

**Comment by villares** [comment]
Some ideas:
- Change the `set_java_home()` function to replace any current JAVA_HOME setting instead of appending a new one.

```python
# proposed change
def set_java_home(jdk_path: pathlib.Path) -> None:
    '''
    Add jdk path to current envirnonment and to Thonny's config file
    as visible in (tools > options > general > env vars)
    '''
    os.environ['JAVA_HOME'] = str(pathlib.Path(THONNY_USER_DIR) / jdk_path)
    jdk_path = 'JAVA_HOME=' + str(jdk_path)
    env_vars = get_workbench().get_option('general.environment')
    # showinfo('debug', env_vars)
    for i, var in enumerate(env_vars):
        if 'JAVA_HOME' in var:
           env_vars[i] = jdk_path
           break
    else:
        env_vars.append(jdk_path)
    # showinfo('debug', env_vars)
    get_workbench().set_option('general.environment', env_vars)
```
- See if Thonny is portable and if JDK-17 is in the expected place call `set_java_home()` to make sure it is always updated... the question is, when?
    - On plug-in load? Will it be soon enough?

UPDATE 1: looks like `env_vars` is a list.. I thought it would be a multiline string...
UPDATE 2: I tried calling `jdk_install_exists()` on plug-in load (afterr the modification above to `set_java_home()`) but it is not soon enough to make `py5`  work. It only sees the `env_vars` modification in the next launch.
UPDATE 3:  Now `set_java_home()` also sets the current envirnoment with `os.environ['JAVA_HOME'] = str(pathlib.Path(THONNY_USER_DIR) / jdk_path)`, as it was doing after the download...  Is this dangerous?

So... this seems to work. Should I check  if this is a portable before I overwrite the JAVA_HOME from the config?
Should I ask the user for permission? Will `set_java_home()` even be called if a valid JAVA_HOME was set for starts? (maybe it won't then we are safe).

UPDATE 4:  I can't wrap my head arround this right now, but I'm sure that with a little tought we could change the following function to only search for the jdk folder, and actually call `set_java_home`, when a valid JAVA_HOME is not already set. And that would be a bit safer, not overriding some user's manual (and valid!) config.

UPDATE 5: Well, I couldn't resist working on it a bit more. See if this makes any sense, pelase:

```python
# proposed change
def jdk_install_exists() -> bool:
    '''check system for existing jdk that meets the py5 version requirements'''
    jdk_dir = 'jdk-' + str(_REQUIRE_JDK)
    system_jdk = 'TBD'
    # check system for existing jdk install (and the version number)
    if os.environ.get('JAVA_HOME') is not None:
        jdk_ver = os.environ['JAVA_HOME'].split('jdk-')[-1].split('.')[0]
        system_jdk = jdk_ver
    # if no JAVA_HOME was set, or if it is not the required version
    if not system_jdk.isdigit() or int(system_jdk) < _REQUIRE_JDK:
        # search for jdk in Thonny's user config directory
        for name in os.listdir(THONNY_USER_DIR):
            if name.startswith(jdk_dir):
                # set JAVA_HOME in Thonny's config with jdk_dir path 
                set_java_home(pathlib.Path(THONNY_USER_DIR) / jdk_dir)
                return True
        # no jdk directory was found
        return False
    else:
        return True
```

**Comment by villares** [comment]
Testing this... it still is some trouble, a useless JAVA_HOME setting remained in Thonny options after moving the folder around... I have to think harder.

UPDATE: I'll try this...

```python
def jdk_install_exists() -> bool:
    '''
    Look for a jdk in Thonny's user directory first, and if one is present,
    set up JAVA_HOME, add it to Thonny's options, and return True. Otherwise,
    look for existing JAVA_HOME pointing to a jdk that meets the py5 version
    requirements and return True/False accordingly.
    '''
    # search for jdk in Thonny's user config directory
    jdk_dir = 'jdk-' + str(_REQUIRE_JDK)
    for name in os.listdir(THONNY_USER_DIR):
        if name.startswith(jdk_dir):
            # set JAVA_HOME in Thonny's config (and in current env) with jdk_dir path 
            set_java_home(pathlib.Path(THONNY_USER_DIR) / jdk_dir)
            return True    
    # check system for existing jdk install (and the version number)
    system_jdk = 'TBD'
    if os.environ.get('JAVA_HOME') is not None:
        jdk_ver = os.environ['JAVA_HOME'].split('jdk-')[-1].split('.')[0]
        system_jdk = jdk_ver
    # False if no JAVA_HOME was set, or if it is not the required version
    return system_jdk.isdigit() and int(system_jdk) >= _REQUIRE_JDK  
```

**Comment by villares** [comment]
@tabreturn I think I have implemented the ideas on this thread in my latest portable experiment, but it was some weeks ago I'm not sure anymore. It would be great if you could test it when you have the time (and if you could validate my assumptions and choices):

https://www.dropbox.com/s/3ue4cx3yf372teg/thonny-4-with-py5-windows-portable.zip?dl=0

**Comment by tabreturn** [comment]
Sorry @villares -- I've been meaning to get back to this. Work has been manic, but I'll try look at this later in the week.

**Comment by villares** [comment]
No hurry! Thanks!
(I should be writing for the PhD...)

**Comment by villares** [comment]
> @tabreturn I think I have implemented the ideas on this thread in my latest portable experiment, but it was some weeks ago I'm not sure anymore. It would be great if you could test it when you have the time (and if you could validate my assumptions and choices):
> 
> https://www.dropbox.com/s/3ue4cx3yf372teg/thonny-4-with-py5-windows-portable.zip?dl=0

Cheers @tabreturn!

Are you planning to do a new release of thonny-py5mode?
I think (I might be wrong) that there are stuff you merged and improved here that are not yet in the PyPI release...

If you could validate the hack I made in the portable version for setting JAVA_HOME even when the portable folder is moved arround, I think this would be something good to incorporate before the next release...

**Comment by tabreturn** [comment]
Just got back to working on the plug-in, @villares. I've started with upgrading `install_jdk`, but I'm facing a few issues with the new version. I'll see how that goes, maybe roll back if I'm not getting anywhere and focus on this issue.

**Reply by villares** [reply]
I hope you don't mind, I created a `windows-portable` branch on the repo to document the changes I made to my portable "distribution" of Thonny + thonny-py5mode + JDK + py5

**Reply by villares** [reply]
Cheers @tabreturn and @hx2A! 

I have worked on the portable version of the plug-in today and I think I might have reached a state that works OK on both portable 
and non-portable Thonny, the code is at https://github.com/tabreturn/thonny-py5mode/compare/main...windows-portable

I hope the jdk_install_exists() and set_java_home() changes are sensible, they mostly tackle a situation where an outdated JAVA_HOME env var is stuck at the Thonny Options... > Environment Variables panel, and also doesn't add duplicated lines to it.

I'd love to have your feedback...

**Reply by villares** [reply]
Dumping more of my thoughts about this...  I kind of just threw in the working code I had in my modified plug-in (for the portable), now, looking carefully, I could refactor it to:
- Extract the Thonny var envs preference setting action from the function checking the JDK
- Finish moving things from `os` to `pathlib.Path` (like iterating through the user_data dir)

Another thorny subject:
- When py5 is installed by the plug-in, it ends up in a "special site packages" location that can't be managed/updated by the user. I find this can be troublesome when a new version of py5 is released, but I really don't know if there is a solution to this. I think users now have to uninstall the plug-in, update py5 and re-install the plug-in...

**Reply by villares** [reply]
Folks, today I saw something that can have confusing consequences for educators and students using the "portable Thonny" solution. I'm not sure if this is a big problem (maybe not) or if there is something we can do about it (probably not) so I'd like at least to document it, share the information with you:

- If the Windows user had previously used non-portable Thonny + py5, it will create `C:\Users\<username>\AppData\Roaming\Python\Python310\site-packages`. It might be an outdated py5 installation, and portable Thonny will look at this first!

- Thonny's GUI package installer will install stuff there too. If you want to install packages "in" the portable (that will be carried along, not left in the user's roaming site-pages), open Thonny's system shell and use `python -m pip install <the-thing-you-want-to-install>`.  

I'll try to investigate this more.

**Reply by hx2A** [reply]
> * When py5 is installed by the plug-in, it ends up in a "special site packages" location that can't be managed/updated by the user. I find this can be troublesome when a new version of py5 is released, but I really don't know if there is a solution to this. I think users now have to uninstall the plug-in, update py5 and re-install the plug-in...

What about updating the plugin? What if updating the plugin forced an update of py5? This isn't that bad of an idea if the plugin is also updated regularly.

**Reply by villares** [reply]
> I hope the jdk_install_exists() and set_java_home() changes are sensible, they mostly tackle a situation where an outdated JAVA_HOME env var is stuck at the Thonny Options... > Environment Variables panel, and also doesn't add duplicated lines to it.

@GoToLoop we don't have to tackle this for this release, but in the future I'd like the plug-in to check if it is running on a Windows portable Thonny (there is an empty flag file on the portable for this)  and make it check & possibly adjust JAVA_HOME  "on launch", because people move the portable folder around. 

We should look into this at some point. My current portable has a slightly hacked plug-in for this but I don't remember what I did exactly. It is something you want it to do on the Windows portable only...

**Comment by villares** [comment]
I'm going to close this, and open a new one, because some stuff already changed, so it will be less confusing for people getting started on this discussion.
